### PR TITLE
[receiver/postgresql] Metric additions

### PR DESCRIPTION
**Pull request by schmikei**
**Description:** Adding these metrics

```
postgresql.bgwriter.buffers.allocated
postgresql.bgwriter.buffers.writes
postgresql.bgwriter.checkpoint.count
postgresql.bgwriter.duration
postgresql.bgwriter.maxwritten.count
postgresql.connections.max
postgresql.database.count
postgresql.index.scans
postgresql.index.size
postgresql.query.block.count
postgresql.query.count
postgresql.query.duration.total
postgresql.query.duration.average
postgresql.replication.data_delay
postgresql.replication.delay
postgresql.table.count
postgresql.table.size
postgresql.vacuum.count
postgresql.wal.age
postgresql.wal.lag
```

**Link to tracking Issue:** #12338

The plan is to merge this PR and then follow up with a more breaking change for adding resource attributes hence there is a couple TODOs to uncomment out that code.

**Testing:** Unit Testing

Also extended integration tests to capture functionality.

**Documentation:** documentation in metadata.yaml more documentation in `receiver/postgresqlreceiver/docs` folder for some of the metrics.

### CODE REVIEW

File: receiver/postgresqlreceiver/README.md [hunk 1]
```diff
@@ -19,16 +19,21 @@ The monitoring user must be granted `SELECT` on `pg_stat_database`.
...
 - `password`
 
 The following settings are optional:
+
 - `endpoint` (default = `localhost:5432`): The endpoint of the postgresql server. Whether using TCP or Unix sockets, this value should be `host:port`. If `transport` is set to `unix`, the endpoint will internally be translated from `host:port` to `/host.s.PGSQL.port`
 - `transport` (default = `tcp`): The transport protocol being used to connect to postgresql. Available options are `tcp` and `unix`.
 
 - `databases` (default = `[]`): The list of databases for which the receiver will attempt to collect statistics. If an empty list is provided, the receiver will attempt to collect statistics for all non-template databases.
 
+- `collect_query_performance` (default = `false`): Whether to collect query performance metrics. This requires enabling the `pg_stat_statements` extension. For more information how to do that, please refer to this [doc](./docs/query-performance.md)
```
Comment (JonathanWamsley) [inline receiver/postgresqlreceiver/README.md:33]: I like how you created a doc detailing this

File: receiver/postgresqlreceiver/docs/query-performance.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Enabling Query Metrics
+
+**Note that the Query Metrics only work on Postgres versions > 10. For more information please refer to the [pg_stat_statements](https://www.postgresql.org/docs/current/pgstatstatements.html) documentation.**
+
+The query metrics (all metrics prefixed with `postgresql.query.*` rely on the `pg_stat_statements` extension to be enabled. To enable them:
+
+For security reasons only super users and members of the `pg_read_all_stats` role are allowed to see the SQL text.
+
+1. On the database issue this create extension command
```
Comment (JonathanWamsley) [inline receiver/postgresqlreceiver/docs/query-performance.md:9]: This is a nit, but change `issue` to `run`

File: receiver/postgresqlreceiver/scraper.go [hunk 3]
```diff
@@ -281,6 +310,162 @@ func (p *postgreSQLScraper) collectBackends(
...
+func (p *postgreSQLScraper) collectBackgroundWriterStats(
+	ctx context.Context,
+	now pcommon.Timestamp,
+	client client,
+	errors scrapererror.ScrapeErrors,
+) {
+	bgStats, err := client.getBackgroundWriterStats(ctx)
+	if err != nil {
+		p.logger.Error("Errors encountered while fetching background writers", zap.Error(err))
+		errors.AddPartial(0, err)
```
Comment (JonathanWamsley) [inline receiver/postgresqlreceiver/scraper.go:381]: Should this be `errors.AddPartial(1, err)` ?

**Review by JonathanWamsley — COMMENTED** [review]
Overall looks very clean! I ran on postgres 10/14 and did not get any errors

**Comment by JonathanWamsley** [comment]
I could not find `connections.open`, all query metrics, `replication.data_delay`, `row.operations` and the `wal.age` and `wal.lag` metrics in the expected integration json files. Are all of these expected to be missing?

**Comment by schmikei** [comment]
> I could not find `connections.open`, all query metrics, `replication.data_delay`, `row.operations` and the `wal.age` and `wal.lag` metrics in the expected integration json files. Are all of these expected to be missing?

Yes they require setting master instances with replica ones to get them, not super easy to do in integration tests but I'll take another look to see if it is really crazy!

**Comment by schmikei** [comment]
@djaglowski do you mind taking a look at this?

**Review by djaglowski — COMMENTED** [review]
This is too many changes for one PR. It's also not clear how this is related to resource attribute changes proposed in #12338.

As mentioned on the issue, we need to break that into multiple issues, and then have smaller more focused PRs. I think roughly: 
1. Fix the way we're collecting to make proper use of resource attributes
2. Add a new resource and associated metrics
3. Next resource and associated metrics
4. ...

File: receiver/postgresqlreceiver/metadata.yaml [hunk 4]
```diff
@@ -1,82 +1,375 @@
 name: postgresqlreceiver
 
+# TODO: use resource attributes
+# resource_attributes:
+#   postgresql.database:
```
[the remark this answers is not in the document]
Comment (schmikei) [reply receiver/postgresqlreceiver/metadata.yaml:5]: @djaglowski okay sounds good! I'll work on separating out in the order that you recommended.

**Comment by schmikei** [comment]
Closing as resolved in other PRs :)

**Pull request closed without merging**